### PR TITLE
[vtv] Add new extractor

### PR DESCRIPTION
**Pull request by tmsbrg**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

VTV is the national television broadcaster of Vietnam. Their website hosts [livestreams](http://us.tvnet.gov.vn/kenh-truyen-hinh/danh-sach) and videos from numerous state-owned television channels. They apparently broadcast worldwide. At least I can see the streams here in the Netherlands. This extractor supports audio and video livestreams as well as static videos.

### CODE REVIEW

File: youtube_dl/extractor/vtv.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            'thumbnail': r're:https?://.*\.png$',
+        }
+
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._html_search_regex(r'<title>(.+?)</title>', webpage, 'title', fatal=False)
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: `default=None`.

File: youtube_dl/extractor/vtv.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            'ext': 'mp4',
+            'title': 'VTV1 - Bắc tuyết tại Lào Cai và Hà Giang - TV Net',
+            'thumbnail': r're:https?://.*\.JPG$',
+        }
+    }, {
+        'url': 'http://us.tvnet.gov.vn/kenh-truyen-hinh/1014',
+        'info_dict': {
+            'id': '1014',
+            'ext': 'm4a',
+            'title': r're:VOV1 | LiveTV - TV Net .*',
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: Any unrelated suffixes should be removed.

File: youtube_dl/extractor/vtv.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        if title is None:
+            title = self._og_search_title(webpage)
+        title.strip()
+
+        mediaplayer_div = self._search_regex(r'(<div[^>]*id="mediaplayer"[^>]*>)', webpage, 'mediaplayer element')
+        mediaplayer_div_attributes = extract_attributes(mediaplayer_div)
+
+        thumbnail = mediaplayer_div_attributes.get("data-image")
+
+        json_url = mediaplayer_div_attributes.get("data-file")
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: Read coding conventions on optional and mandatory fields.

File: youtube_dl/extractor/vtv.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        if title.startswith("VOV"):
+            ext = "m4a"
+        else:
+            ext = "mp4"
+
+        # get any working playlist from streams. Currently there's 2 and the first always works,
+        # but you never know in the future
+        for stream in video_streams:
+            formats = self._extract_m3u8_formats(stream.get("url"), video_id, ext=ext, fatal=False)
+            if len(formats) != 0:
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: `not formats`.

File: youtube_dl/extractor/vtv.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                break
+
+        if re.match(r'https?://[^/]*/video/.*', url) is not None:
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: `/video/ in url`.

File: youtube_dl/extractor/vtv.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            is_live = False
+        elif re.match(r'https?://[^/]*/kenh-truyen-hinh/.*', url) is not None:
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: `/kenh-truyen-hinh/ in url`.

File: youtube_dl/extractor/vtv.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        mediaplayer_div_attributes = extract_attributes(mediaplayer_div)
+
+        thumbnail = mediaplayer_div_attributes.get("data-image")
+
+        json_url = mediaplayer_div_attributes.get("data-file")
+        video_streams = self._download_json(json_url, video_id)
+
+        # little hack to better support radio streams
+        if title.startswith("VOV"):
+            ext = "m4a"
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: `vcodec` of format should be set to `'none'`.

File: youtube_dl/extractor/vtv.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+from .common import InfoExtractor
+
+import re
+
+from ..utils import extract_attributes
+
+class VTVIE(InfoExtractor):
+    _VALID_URL = r'https?://..\.tvnet\.gov\.vn/[^/]*/(?P<id>[0-9]+)/?.*'
+    _TESTS = [{
+        'url': 'http://us.tvnet.gov.vn/kenh-truyen-hinh/1011/vtv1',
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: Each test should be commented on what it actually tests.

File: youtube_dl/extractor/vtv.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+import re
+
+from ..utils import extract_attributes
+
+class VTVIE(InfoExtractor):
+    _VALID_URL = r'https?://..\.tvnet\.gov\.vn/[^/]*/(?P<id>[0-9]+)/?.*'
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: No `..`. `.*` at the end is pointless.

**Comment by tmsbrg** [comment]
Thanks for the review! Hope this commit addresses the issues you found correctly.

**Comment by dstftw** [comment]
>- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

Now actually do that.

File: youtube_dl/extractor/vtv.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+import re
+
+from ..utils import extract_attributes
+
+class VTVIE(InfoExtractor):
+    _VALID_URL = r'https?://(au|ca|cz|de|jp|kr|tw|us|vn)\.tvnet\.gov\.vn/[^/]*/(?P<id>[0-9]+)/?'
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: Don't capture groups you don't use. Use proper regex to match all country codes.

File: youtube_dl/extractor/vtv.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+
+import re
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: Generic imports should go before youtube-dl's.

File: youtube_dl/extractor/vtv.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+        title = self._html_search_regex(r'<title>(.+?)</title>', webpage, 'title', default=None, fatal=False)
+        if title is None:
+            title = self._og_search_title(webpage)
+        title.strip()
+
+        mediaplayer_div = self._search_regex(r'(<div[^>]*id="mediaplayer"[^>]*>)', webpage, 'mediaplayer element')
+        mediaplayer_div_attributes = extract_attributes(mediaplayer_div)
+
+        thumbnail = mediaplayer_div_attributes.get("data-image")
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: Single quotes.

File: youtube_dl/extractor/vtv.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+        json_url = mediaplayer_div_attributes["data-file"]
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: Same.

File: youtube_dl/extractor/vtv.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+            is_live = None
+
+        if is_live:
+            title = self._live_title(title)
+
+        return {
+            'id': video_id,
+            'title': title,
+            'thumbnail': thumbnail,
+            'formats': formats,
```
Comment (dstftw) [inline youtube_dl/extractor/vtv.py]: Formats not sorted.

File: youtube_dl/extractor/vtv.py (hunk 10 above)
Comment (tmsbrg) [reply youtube_dl/extractor/vtv.py]: These are the only country codes that are valid at the moment. Do you think it'd be better to expand it to capture every country code, or even [a-z][a-z]?

**Pull request closed without merging**